mirgecom/steppers.py

```diff
@@ -0,0 +1,184 @@
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+import logging
+import numpy as np
+import numpy.linalg as la  # noqa
+import pyopencl as cl
+import pyopencl.array as cla  # noqa
+
+from grudge.eager import EagerDGDiscretization
+from grudge.shortcuts import make_visualizer
+from mirgecom.euler import get_inviscid_timestep
+from mirgecom.euler import inviscid_operator
+from mirgecom.euler import split_fields
+from mirgecom.integrators import rk4_step
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+
+
+def advance_state(rhs, timestepper, checkpoint, get_timestep,
+                  state, t=0.0, t_final=1.0, istep=0):
+    """
+    Implements a generic state advancement routine
+    """
+    if t_final <= t:
+        return(istep, t, state)
+
+    while t < t_final:
+
+        dt = get_timestep(state=state)
+        if dt < 0:
+            return (istep, t, state)
+
+        status = checkpoint(state=state, step=istep, t=t, dt=dt)
+        if status != 0:
+            return (istep, t, state)
+
+        state = timestepper(state, t, dt, rhs)
+
+        t += dt
+        istep += 1
+
+    return (istep, t, state)
+
+
+def euler_flow_stepper(parameters, ctx_factory=cl.create_some_context):
+    """
+    Deprecated: Implements a generic time stepping loop for an inviscid flow.
+    """
+    cl_ctx = ctx_factory()
+    queue = cl.CommandQueue(cl_ctx)
```

Comment:
Shouldn't create command queues anywhere other than the top level. (Follow-up from #13)